{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Secondary analysis of 10x visium data post-alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clusters with scVI \n",
    "import os\n",
    "import scvi\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "\n",
    "os.chdir('/Users/amanda.mitchell/Dropbox/10x_visium_paper_2024/data/expression-data/sagittal')\n",
    "adata = ad.read_h5ad('/Users/amanda.mitchell/Dropbox/10x_visium_paper_2024/data/expression-data/sagittal/sagittal_vis.h5ad')\n",
    "\n",
    "# normalize data\n",
    "sc.pp.filter_cells(adata, min_genes=200)\n",
    "sc.pp.filter_genes(adata, min_cells=3)\n",
    "adata.layers[\"counts\"] = adata.X.copy()  # preserve counts\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "adata.raw = adata\n",
    "sc.pp.highly_variable_genes(adata, n_top_genes=600, subset=True)\n",
    "\n",
    "# pca embedding\n",
    "sc.pp.pca(adata, n_comps=50)\n",
    "adata.write_h5ad('/Users/amanda.mitchell/Desktop/sagittal_vis_pca.h5ad')\n",
    "\n",
    "# scVI embedding\n",
    "scvi.model.SCVI.setup_anndata(adata, layer=\"counts\",\n",
    "    categorical_covariate_keys=[\"sample\"]) #,continuous_covariate_keys=[\"percent_mito\", \"percent_ribo\"])\n",
    "model = scvi.model.SCVI(adata)\n",
    "model.train()\n",
    "latent = model.get_latent_representation()\n",
    "adata.obsm[\"X_scVI\"] = latent\n",
    "adata.layers[\"scvi_normalized\"] = model.get_normalized_expression(library_size=10e4)\n",
    "adata.write_h5ad('/Users/amanda.mitchell/Desktop/sagittal_vis_scvi.h5ad')\n",
    "\n",
    "# plot model metrics\n",
    "# model.history()\n",
    "# visualization with scVI embedding\n",
    "#sc.pp.neighbors(adata, use_rep=\"X_scVI\")\n",
    "sc.pp.neighbors(adata, use_rep=\"X_pca\")\n",
    "sc.tl.umap(adata, min_dist=0.3)\n",
    "sc.pl.umap(adata, color=[\"WB_subclass\"], frameon=False)\n",
    "\n",
    "# clustering\n",
    "sc.tl.leiden(adata, key_added=\"leiden_pca\", resolution=0.8)\n",
    "sc.pl.umap(adata, color=[\"leiden_pca\"], frameon=False)\n",
    "# sc.tl.leiden(adata, key_added=\"leiden_scVI\", resolution=0.8)\n",
    "# sc.pl.umap(adata, color=[\"leiden_scVI\"], frameon=False)\n",
    "adata.write_h5ad('/Users/amanda.mitchell/Desktop/sagittal_vis_analyzed.h5ad')\n",
    "\n",
    "# differential expression\n",
    "adata.obs.leiden_pca.head()\n",
    "de_df = model.differential_expression(groupby=\"WB_subclass\")\n",
    "de_df.head()\n",
    "\n",
    "# markers\n",
    "markers = {}\n",
    "cats = adata.obs.cell_type.cat.categories\n",
    "for i, c in enumerate(cats):\n",
    "    cid = f\"{c} vs Rest\"\n",
    "    leiden_pca_df = de_df.loc[de_df.comparison == cid]\n",
    "    leiden_pca_df = leiden_pca_df[leiden_pca_df.lfc_mean > 0]\n",
    "    leiden_pca_df = leiden_pca_df[leiden_pca_df[\"bayes_factor\"] > 3]\n",
    "    leiden_pca_df = leiden_pca_df[leiden_pca_df[\"non_zeros_proportion1\"] > 0.1]\n",
    "    markers[c] = leiden_pca_df.index.tolist()[:5]\n",
    "\n",
    "sc.tl.dendrogram(adata, groupby=\"leiden_pca\", use_rep=\"X_pca\")\n",
    "sc.pl.dotplot(adata, markers, groupby=\"leiden_pca\", dendrogram=True,\n",
    "    color_map=\"Blues\", swap_axes=True, use_raw=True, standard_scale=\"var\")\n",
    "\n",
    "dotplot.savefig(\"/Users/amanda.mitchell/Dropbox/10x_visium_paper_2024/figures/leiden_markers_dotplot.png\", dpi=300, bbox_inches='tight')  # Save as PNG\n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cNMF modules with sklearn\n",
    "from sklearn.decomposition import NMF\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anndata as ad\n",
    "import os\n",
    "\n",
    "os.chdir('/Users/amanda.mitchell/Dropbox/10x_visium_paper_2024/data/expression-data/sagittal')\n",
    "adata = ad.read_h5ad('/Users/amanda.mitchell/Dropbox/10x_visium_paper_2024/data/expression-data/sagittal/sagittal_vis.h5ad')\n",
    "#sample_size = int(adata.n_obs * 0.01)\n",
    "#random_indices = np.random.choice(adata.n_obs, size=sample_size, replace=False)\n",
    "#adata_sample = adata[random_indices, :]\n",
    "#counts = adata_sample.to_df()\n",
    "matrix = adata.to_df()\n",
    "model = NMF(n_components=20, init='random', random_state=0)\n",
    "W = pd.DataFrame(model.fit_transform(matrix)) # samples by components\n",
    "H = pd.DataFrame(model.components_) # components x genes\n",
    "W.to_csv('/Users/amanda.mitchell/Desktop/sagittal_vis_W.csv')\n",
    "H.to_csv('/Users/amanda.mitchell/Desktop/sagittal_vis_H.csv')\n",
    "\n",
    "max_W = W.idxmax(axis=1)\n",
    "max_W.index = adata.obs.index\n",
    "adata.obs['module']=pd.DataFrame(max_W)\n",
    "\n",
    "max_H = H.idxmax(axis=0)\n",
    "max_H.index = adata.var.index\n",
    "adata.var['program']=pd.DataFrame(max_H)\n",
    "adata.write_h5ad('/Users/amanda.mitchell/Desktop/sagittal_vis_cNMF.h5ad')\n",
    "\n",
    "\n",
    "## module markers\n",
    "# differential expression\n",
    "adata.obs.module.head()\n",
    "de_df = model.differential_expression(groupby=\"module\")\n",
    "de_df.head()\n",
    "\n",
    "# markers\n",
    "markers = {}\n",
    "cats = adata.obs.module.cat.categories\n",
    "for i, c in enumerate(cats):\n",
    "    cid = f\"{c} vs Rest\"\n",
    "    module_df = de_df.loc[de_df.comparison == cid]\n",
    "    module_df = module_df[module_df.lfc_mean > 0]\n",
    "    module_df = module_df[module_df[\"bayes_factor\"] > 3]\n",
    "    module_df = module_df[module_df[\"non_zeros_proportion1\"] > 0.1]\n",
    "    markers[c] = module_df.index.tolist()[:5]\n",
    "\n",
    "#sc.tl.dendrogram(adata, groupby=\"module\", use_rep=\"X_pca\")\n",
    "sc.pl.dotplot(adata, markers, groupby=\"module\", dendrogram=False,\n",
    "    color_map=\"Blues\", swap_axes=True, use_raw=True, standard_scale=\"var\")\n",
    " \n",
    "dotplot.savefig(\"/Users/amanda.mitchell/Dropbox/10x_visium_paper_2024/figures/module_markers_dotplot.png\", dpi=300, bbox_inches='tight')  # Save as PNG\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# integration with scVI\n",
    "\n",
    "from sklearn.decomposition import NMF\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import scvi\n",
    "import os\n",
    "\n",
    "os.chdir('/Users/amanda.mitchell/Dropbox/10x_visium_paper_2024/data/expression-data/sagittal')\n",
    "adata1 = ad.read_h5ad('/Users/amanda.mitchell/Dropbox/10x_visium_paper_2024/data/expression-data/sagittal/sagittal_vis.h5ad')\n",
    "sample_size1 = int(adata1.n_obs * 0.01)\n",
    "random_indices1 = np.random.choice(adata1.n_obs, size=sample_size1, replace=False)\n",
    "adata1s = adata1[random_indices1, :]\n",
    "\n",
    "sample_size2 = int(adata1.n_obs * 0.01)\n",
    "random_indices2 = np.random.choice(adata1.n_obs, size=sample_size1, replace=False)\n",
    "adata2s = adata1[random_indices2, :]\n",
    "\n",
    "adata1s.obs['dataset']='1s'\n",
    "adata2s.obs['dataset']='2s'\n",
    "adatas = ad.concat([adata1s, adata2s])\n",
    "\n",
    "scvi.model.SCVI.setup_anndata(adatas, batch_key=\"dataset\")\n",
    "model = scvi.model.SCVI(adatas, n_layers=2, n_latent=30, gene_likelihood=\"nb\")\n",
    "model.train()\n",
    "SCVI_LATENT_KEY = \"X_scVI\"\n",
    "adatas.obsm[SCVI_LATENT_KEY] = model.get_latent_representation()\n",
    "sc.pp.neighbors(adatas, use_rep=SCVI_LATENT_KEY)\n",
    "sc.tl.leiden(adatas)\n",
    "SCVI_MDE_KEY = \"X_scVI_MDE\"\n",
    "sc.pl.embedding(adatas, basis=SCVI_MDE_KEY, color=[\"dataset\"], frameon=False, ncols=1)\n",
    "embedding.savefig(\"/Users/amanda.mitchell/Dropbox/10x_visium_paper_2024/figures/dataset_scvit.png\", dpi=300)\n",
    "\n",
    "# markers\n",
    "markers = {}\n",
    "cats = adatas.obs.leiden.cat.categories\n",
    "for i, c in enumerate(cats):\n",
    "    cid = f\"{c} vs Rest\"\n",
    "    cell_type_df = de_df.loc[de_df.comparison == cid]\n",
    "    cell_type_df = cell_type_df[cell_type_df.lfc_mean > 0]\n",
    "    cell_type_df = cell_type_df[cell_type_df[\"bayes_factor\"] > 3]\n",
    "    cell_type_df = cell_type_df[cell_type_df[\"non_zeros_proportion1\"] > 0.1]\n",
    "    markers[c] = cell_type_df.index.tolist()[:3]\n",
    "\n",
    "sc.tl.dendrogram(adatas, groupby=\"leiden\", use_rep=\"X_scVI\")\n",
    "sc.pl.dotplot(adatas, markers, groupby=\"leiden\", dendrogram=True,\n",
    "    color_map=\"Blues\", swap_axes=True, use_raw=True, standard_scale=\"var\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# integration wtih scANVI\n",
    "\n",
    "scanvi_model = scvi.model.SCANVI.from_scvi_model(\n",
    "    model,\n",
    "    adata=adata,\n",
    "    labels_key=\"cell_type\",\n",
    "    unlabeled_category=\"Unknown\",\n",
    ")\n",
    "scanvi_model.train(max_epochs=20, n_samples_per_label=100)\n",
    "SCANVI_LATENT_KEY = \"X_scANVI\"\n",
    "adata.obsm[SCANVI_LATENT_KEY] = scanvi_model.get_latent_representation(adata)\n",
    "SCANVI_MDE_KEY = \"X_scANVI_MDE\"\n",
    "adata.obsm[SCANVI_MDE_KEY] = scvi.model.utils.mde(adata.obsm[SCANVI_LATENT_KEY], accelerator=\"cpu\")\n",
    "sc.pl.embedding(adata, basis=SCANVI_MDE_KEY, color=[\"cell_type\"], ncols=1, frameon=False)\n",
    "\n",
    "# integration metrics\n",
    "bm = Benchmarker(\n",
    "    adata,\n",
    "    batch_key=\"batch\",\n",
    "    label_key=\"cell_type\",\n",
    "    embedding_obsm_keys=[\"X_pca\", SCVI_LATENT_KEY, SCANVI_LATENT_KEY],\n",
    "    n_jobs=-1,\n",
    ")\n",
    "bm.benchmark()\n",
    "bm.plot_results_table(min_max_scale=False)\n",
    "df = bm.get_results(min_max_scale=False)\n",
    "print(df)\n",
    "\n",
    "\n",
    "# markers\n",
    "markers = {}\n",
    "cats = adata.obs.cell_type.cat.categories\n",
    "for i, c in enumerate(cats):\n",
    "    cid = f\"{c} vs Rest\"\n",
    "    cell_type_df = de_df.loc[de_df.comparison == cid]\n",
    "    cell_type_df = cell_type_df[cell_type_df.lfc_mean > 0]\n",
    "    cell_type_df = cell_type_df[cell_type_df[\"bayes_factor\"] > 3]\n",
    "    cell_type_df = cell_type_df[cell_type_df[\"non_zeros_proportion1\"] > 0.1]\n",
    "    markers[c] = cell_type_df.index.tolist()[:3]\n",
    "\n",
    "sc.tl.dendrogram(adata, groupby=\"WB_subclass\", use_rep=\"X_scVI\")\n",
    "sc.pl.dotplot(adata, markers, groupby=\"WB_subclass\", dendrogram=True,\n",
    "    color_map=\"Blues\", swap_axes=True, use_raw=True, standard_scale=\"var\")\n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# graphs of clusters and modules\n",
    "\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "component_df = pd.get_dummies(adata.obs['component'])\n",
    "WB_subclass_df = pd.get_dummies(adata.obs['WB_subclass'])\n",
    "df = pd.concat([component_df, WB_subclass_df], axis=1)\n",
    "\n",
    "sns.clustermap(component_df, cmap=\"viridis\", figsize=(10,8), metric='euclideam', method=\"average\", linewidths=0.5)\n",
    "\n",
    "clustermap.ax_heatmap.set_title(\"Clustered Heatmap of Components vs Cells\")\n",
    "clustermap.ax_heatmap.set_xlabel(\"Components\")\n",
    "clustermap.ax_heatmap.set_ylabel(\"Cells\")\n",
    "\n",
    "# Save the clustered heatmap as a PNG file\n",
    "clustered_filename = \"/Users/amanda.mitchell/Dropbox/10x_visium_paper_2024/figures/clustered_modules_clusters.png\"\n",
    "plt.savefig(clustered_filename)\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
