{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d61c97f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Set libraries and read in data\n",
    "\n",
    "#includes xx libraries: specify library\n",
    "#change file locations \n",
    "\n",
    "#pip install if needed\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from scipy import stats\n",
    "import anndata as ad\n",
    "from anndata import AnnData\n",
    "import configome as co\n",
    "import harmonypy as hp\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scdrs\n",
    "import besca as bc\n",
    "import scanpy.external as sce\n",
    "\n",
    "adata = sc.read_h5ad('/mnt/ps/home/CORP/storage/tenx/runa/scanpy/h5ad_files/processed/celltype.h5ad') \n",
    "sc.settings.verbosity = 3 # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')\n",
    "results_file = '/mnt/ps/home/CORP/storage/tenx/runa/scanpy/h5ad_files/processed/celltype.h5ad'\n",
    "results_file_raw = '/mnt/ps/home/CORP/storage/tenx/runa/scanpy/h5ad_files/raw/celltype.h5ad'\n",
    "## gs://comp-neuro/ #CompNeuro Google Cloud bucket\n",
    "\n",
    "# RXRX samples\n",
    "mysamples = \"/mnt/bh1/data/storage/outs/\"\n",
    "myclusters = \"/mnt/ps/home/CORP/storage/tenx/runa/scanpy/celltype\"\n",
    "os.chdir(mysamples)\n",
    "\n",
    "adata = sc.read_10x_h5('filtered_feature_bc_matrix.h5', gex_only = True) \n",
    "adata.var_names_make_unique()\n",
    "adata.write(results_file, compression='gzip')\n",
    "adata.write(results_file_raw, compression='gzip')\n",
    "os.chdir(myclusters)\n",
    "\n",
    "## Add scDRS, disease relevance score ##\n",
    "DATA_PATH = scdrs.__path__[0]\n",
    "H5AD_FILE = os.path.join(DATA_PATH, '/mnt/ps/home/CORP/storage/tenx/runa/scanpy/h5ad_files/processed/celltype.h5ad')\n",
    "GS_FILE = os.path.join(DATA_PATH, \"/mnt/ps/home/CORP/amanda.mitchell/data/data/external/genetic-datasets/GWAS/magma_10kb_1000.74_traits.gs\")\n",
    "\n",
    "#Location of comp-neuro Google bucket gs://comp-neuro/\n",
    "\n",
    "#Marker genes\n",
    "ttestm = \"celltype_markersttest.csv\"\n",
    "wilcm = \"celltype_markerswilcoxon.csv\"\n",
    "ctwilcm = \"celltype_markersttest.csv\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1546882f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "adata.write(results_file_raw, compression='gzip')\n",
    "adata.write(results_file, compression='gzip')\n",
    "os.chdir(myclusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f291a8aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Preprocess and cluster\n",
    "\n",
    "sc.pl.highest_expr_genes(adata, n_top=30, save='.png')\n",
    "sc.pp.filter_cells(adata, min_genes=200)\n",
    "#sc.pp.filter_genes(adata, min_cells=3)\n",
    "adata.var['mt'] = adata.var_names.str.startswith('MT-') \n",
    "\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "sc.pl.violin(adata, ['n_genes_by_counts', 'total_counts', 'pct_counts_mt'], jitter=0.4, multi_panel=True, save='QCs.png')\n",
    "adata = adata[adata.obs.n_genes_by_counts > 1000, :]\n",
    "adata = adata[adata.obs.pct_counts_mt < 15, :]\n",
    "sc.pl.violin(adata, ['n_genes_by_counts', 'total_counts', 'pct_counts_mt'], jitter=0.4, multi_panel=True, save='QCs_post.png')\n",
    "\n",
    "## Normalize, log, variable genes ##\n",
    "sc.pp.normalize_total(adata, target_sum=1e6) #CPM 1e6, by cells 1e4, len(adata), exclude_highly_expressed=True\n",
    "sc.pp.log1p(adata)\n",
    "#sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5) #how to change\n",
    "sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5, batch_key = 'batch')\n",
    "sc.pl.highly_variable_genes(adata, save='.png')\n",
    "adata.raw = adata\n",
    "\n",
    "## Filter ##\n",
    "#adata.2 = adata[:, adata.var.highly_variable]\n",
    "sc.pp.regress_out(adata, ['total_counts', 'pct_counts_mt']) #may overcorrect\n",
    "sc.pp.scale(adata)#, max_value=10)\n",
    "\n",
    "# PCA\n",
    "sc.tl.pca(adata, svd_solver='arpack')\n",
    "sc.pl.pca(adata, save='.png')\n",
    "adata.write(results_file)\n",
    "adata\n",
    "\n",
    "## Cluster ##\n",
    "sc.pl.pca_variance_ratio(adata, log=True, n_pcs = 50, save='elbow.png')\n",
    "sc.pp.neighbors(adata, n_neighbors=10, n_pcs=30) #change to minimal PCs based on Elbow\n",
    "sc.tl.leiden(adata, resolution=1, n_iterations=100)\n",
    "\n",
    "## Batch correction\n",
    "adatab=adata\n",
    "sce.pp.harmony_integrate(adatab, 'batch')\n",
    "sc.pl.embedding(adatab, basis='X_pca_harmony', color=['batch'], save='batch_correction.png')\n",
    "adatab.obsm['X_pca'] = adatab.obsm['X_pca_harmony']\n",
    "sc.pp.neighbors(adatab, n_neighbors=10, n_pcs=30) #change to minimal PCs based on Elbow\n",
    "sc.tl.leiden(adatab, resolution=1, n_iterations=100)\n",
    "\n",
    "#UMAP\n",
    "sc.tl.paga(adata)\n",
    "sc.pl.paga(adata, plot=False)  \n",
    "sc.tl.umap(adata, init_pos='paga')\n",
    "sc.pl.umap(adata, color=['leiden'], save='leiden.png')\n",
    "sc.pl.umap(adata, color=['batch'], save='leiden-batch.png')\n",
    "\n",
    "sc.tl.paga(adatab)\n",
    "sc.pl.paga(adatab, plot=False)  \n",
    "sc.tl.umap(adatab, init_pos='paga')\n",
    "sc.pl.umap(adatab, color=['leiden'], save='corrected-leiden.png')\n",
    "sc.pl.umap(adatab, color=['batch'], save='corrected-leiden-batch.png')\n",
    "\n",
    "broad_class = [\"SLC17A6\", \"GAD2\", \"GFAP\", \"SOX2\", \"POU5F1\", \"NKX2-1\", \"NEUROG2\", \"TH\"]\n",
    "THsub = ['TH', \"NR4A2\", 'SLC6A3', 'GAD2']\n",
    "mNPC = [\"LMX1B\", \"FOXA2\", \"NR4A2\", \"NES\"]\n",
    "exc = [\"SLC17A6\", \"SLC17A7\", \"SLC7A6\", \"SLC7A7\"]\n",
    "inh = [\"GAD1\", \"GAD2\", \"SLC6A1\", \"GPHN\"]\n",
    "\n",
    "sc.pl.umap(adata, color=broad_class, use_raw=True, save='broad_class.png')\n",
    "sc.pl.umap(adata, color=THsub, use_raw=True, save='THsub.png')\n",
    "sc.pl.umap(adata, color=mNPC, use_raw=True, save='midbrainNPC_ct.png')\n",
    "sc.pl.umap(adata, color=exc, use_raw=True, save='excitatory_umap.png')\n",
    "sc.pl.umap(adata, color=inh, use_raw=True, save='inhibitory_umap.png')\n",
    "\n",
    "sc.pl.dotplot(adata, broad_class, groupby='leiden', use_raw=True, save='broad_class.png')\n",
    "sc.pl.dotplot(adata, THsub, groupby='leiden', use_raw=True, save='THsub.png')\n",
    "sc.pl.dotplot(adata, iMN, groupby='leiden', use_raw=True, save='iMN.png')\n",
    "sc.pl.dotplot(adata, mNPC, groupby='leiden', use_raw=True, save='mNPC.png')\n",
    "sc.pl.dotplot(adata, exc, groupby='leiden', use_raw=True, save='exc.png')\n",
    "sc.pl.dotplot(adata, inh, groupby='leiden', use_raw=True, save='inh.png')\n",
    "\n",
    "adata.write(results_file)\n",
    "adata.write(results_file_corrected)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4920fe03",
   "metadata": {},
   "outputs": [],
   "source": [
    "## QCs on UMAP\n",
    "\n",
    "sc.pl.umap(adata, color=['n_genes_by_counts'], save='genecounts.png')\n",
    "sc.pl.umap(adata, color=['pct_counts_mt'], save='pct_counts_mt.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ccc147b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Graph key genes and determine marker genes per cluster\n",
    "\n",
    "#paper_iPSC = [\"CNMD\", \"SFRP2\", \"PDLIM2\", \"CTSB\"]\n",
    "#paper_pluri = [\"POU5F1\", \"SOX2\", \"SOX4\", \"PAX6\", \"FUT4\", \"NANOG\"]\n",
    "\n",
    "neuronal = [\"TUBB3\",\"MAP2\",\"DLG4\", \"SYT1\", \"SYT7\", \"SYN1\", \"BSN\", 'STXBP1']\n",
    "subtypes = ['SLC6A4', 'CHAT', 'DRD1', 'DRD2', 'PPP1R1B']\n",
    "iMN = [\"ISL1\", \"MNX1\", \"FOXP1\", \"OLIG2\", \"COL3A1\", \"TAGLN\", \"SOX2\", \"TOP2A\", \"STMN2\", \"ONECUT2\", \"PHOX2A\"]\n",
    "iDA = [\"TH\", \"TPH1\", \"SYT4\", \"SCN3A\", \"FOXA2\", \"LMX1A\", \"LMX1B\", \"NR4A2\"]\n",
    "uglia_a = [\"TMEM119\", \"HEXB\", \"TREM2\", \"AIF1\", \"ARHGAP5\", \"CCR5\", \"CD164\", \"CSMD3\", \"CST3\", \"CX3CR1\", \"GCNT1\", \"GOLM1\"]\n",
    "uglia_b = [\"GPR155\", \"GPR34\", \"ADGRG1\", \"GTF2H2\", \"LRBA\", \"LRRC3\", \"MED12L\", \"MFAP3\", \n",
    "          \"P2RY12\", \"P2RY13\", \"PLXDC2\", \"PMEPA1\", \"RAB39A\", \"SPARC\", \"TLR3\", \"AQP4\"]\n",
    "glu = [\"SLC17A6\", \"SLC17A7\", \"KMT5C\", \"PTEN\", \"CHD2\", \"TCF4\", \"RELN\", \"SCN1A\", \"GRIN2B\", \"FOXP1\", \"STXBP1\"]\n",
    "\n",
    "sc.pl.umap(adata, color=neuronal, use_raw=True, save='neuronal_ct.png')\n",
    "sc.pl.umap(adata, color=subtypes, use_raw=True, save='neuronalsubtypes_ct.png')\n",
    "sc.pl.umap(adata, color=iMN, use_raw=True, save='iMN_ct.png')\n",
    "sc.pl.umap(adata, color=iDA, use_raw=True, save='iDA_ct.png')\n",
    "sc.pl.umap(adata, color=uglia_a, use_raw=True, save='iuglia_a_ct.png')\n",
    "sc.pl.umap(adata, color=uglia_b, use_raw=True, save='iuglia_b_ct.png')\n",
    "sc.pl.umap(adata, color=glu, use_raw=True, save='glu_ct.png')\n",
    "\n",
    "# Marker genes, test\n",
    "sc.tl.rank_genes_groups(adata, 'leiden', method='t-test')\n",
    "sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False, save='ttest.png')\n",
    "sc.settings.verbosity = 2  # reduce the verbosity\n",
    "sc.pl.rank_genes_groups_matrixplot(adata, n_genes=3, standard_scale='var', cmap='Blues', save='ttest_mg_matrixplot.png')\n",
    "sc.pl.rank_genes_groups_dotplot(adata, n_genes=3, standard_scale='var',save='ttest_mg_dotplot.png')\n",
    "DEttest=pd.DataFrame(adata.uns['rank_genes_groups']['names']).head(10)\n",
    "DEttest.columns ='leiden_' + DEttest.columns\n",
    "DEttest.to_csv(ttestm)\n",
    "\n",
    "# Marker genes, wilcoxon\n",
    "sc.tl.rank_genes_groups(adata, 'leiden', method='wilcoxon')\n",
    "sc.pl.rank_genes_groups_dotplot(adata, n_genes=4, save='mg_dotplot.png')\n",
    "sc.pl.rank_genes_groups_matrixplot(adata, n_genes=3, standard_scale='var', cmap='Blues', save='wilcoxon_mg_matrixplot.png')\n",
    "sc.pl.rank_genes_groups_dotplot(adata, n_genes=3, standard_scale='var', cmap='Blues', save='wilcoxon_mg_dotplot.png')\n",
    "DEwil=pd.DataFrame(adata.uns['rank_genes_groups']['names']).head(10)\n",
    "DEwil.columns ='leiden_' + DEwil.columns\n",
    "DEwil.to_csv(wilcm)\n",
    "\n",
    "## Dotplots ##\n",
    "#sc.pl.stacked_violin(adata, neuronal, groupby='leiden', rotation=90, save='ranked_neuronal.png')\n",
    "sc.pl.dotplot(adata, neuronal, groupby='leiden', save='neuronal.png')\n",
    "sc.pl.dotplot(adata, subtypes, groupby='leiden', save='neuronalsubtypes.png')\n",
    "sc.pl.dotplot(adata, iMN, groupby='leiden', save='iMN.png')\n",
    "sc.pl.dotplot(adata, iDA, groupby='leiden', save='iDA.png')\n",
    "sc.pl.dotplot(adata, uglia_a, groupby='leiden', save='uglia_a.png')\n",
    "sc.pl.dotplot(adata, uglia_b, groupby='leiden', save='uglia_b.png')\n",
    "sc.pl.dotplot(adata, glu, groupby='leiden', save='glu.png')\n",
    "\n",
    "adata.write(results_file, compression='gzip')\n",
    "#adata.raw.to_adata().write('./write/BatchE.h5ad')\n",
    "#adata.uns['rank_genes_groups']\n",
    "\n",
    "adata.uns['rank_genes_groups'].to_csv(./clustermarkers.csv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b04b54c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Single cell disease risk score, scRDS\n",
    "\n",
    "# Load .h5ad file, .cov file, and .gs file\n",
    "#adata = scdrs.util.load_h5ad(H5AD_FILE, flag_filter_data=False, flag_raw_count=False)\n",
    "df_gs = scdrs.util.load_gs(GS_FILE)\n",
    "\n",
    "#Preproecssing .h5ad data compute scDRS score\n",
    "scdrs.preprocess(adata) #preprocess\n",
    "\n",
    "gene_list_SWB = df_gs['PASS_SWB'][0]\n",
    "gene_weight_SWB = df_gs['PASS_SWB'][1]\n",
    "df_res_SWB = scdrs.score_cell(adata, gene_list_SWB, gene_weight=gene_weight_SWB, n_ctrl=20)\n",
    "\n",
    "gene_list_MDD = df_gs['PASS_MDD_Howard2019'][0]\n",
    "gene_weight_MDD = df_gs['PASS_MDD_Howard2019'][1]\n",
    "df_res_MDD = scdrs.score_cell(adata, gene_list_MDD, gene_weight=gene_weight_MDD, n_ctrl=20)\n",
    "\n",
    "gene_list_AD = df_gs['PASS_Alzheimers_Jansen2019'][0]\n",
    "gene_weight_AD = df_gs['PASS_Alzheimers_Jansen2019'][1]\n",
    "df_res_AD = scdrs.score_cell(adata, gene_list_AD, gene_weight=gene_weight_AD, n_ctrl=20)\n",
    "\n",
    "gene_list_Worry = df_gs['PASS_Worry_Nagel2018'][0]\n",
    "gene_weight_Worry = df_gs['PASS_Worry_Nagel2018'][1]\n",
    "df_res_Worry = scdrs.score_cell(adata, gene_list_Worry, gene_weight=gene_weight_Worry, n_ctrl=20)\n",
    "\n",
    "gene_list_EDU_COLLEGE = df_gs['UKB_460K.cov_EDU_COLLEGE'][0]\n",
    "gene_weight_EDU_COLLEGE = df_gs['UKB_460K.cov_EDU_COLLEGE'][1]\n",
    "df_res_EDU_COLLEGE = scdrs.score_cell(adata, gene_list_EDU_COLLEGE, gene_weight=gene_weight_EDU_COLLEGE, n_ctrl=20)\n",
    "\n",
    "gene_list_EDU_YEARS = df_gs['UKB_460K.cov_EDU_YEARS'][0]\n",
    "gene_weight_EDU_YEARS = df_gs['UKB_460K.cov_EDU_YEARS'][1]\n",
    "df_res_EDU_YEARS = scdrs.score_cell(adata, gene_list_EDU_YEARS, gene_weight=gene_weight_EDU_YEARS, n_ctrl=20)\n",
    "\n",
    "gene_list_Verbal = df_gs['PASS_VerbalNumericReasoning_Davies2018'][0]\n",
    "gene_weight_Verbal = df_gs['PASS_VerbalNumericReasoning_Davies2018'][1]\n",
    "df_res_Verbal = scdrs.score_cell(adata, gene_list_Verbal, gene_weight=gene_weight_Verbal, n_ctrl=20)\n",
    "\n",
    "gene_list_Intelligence = df_gs['PASS_Intelligence_SavageJansen2018'][0]\n",
    "gene_weight_Intelligence = df_gs['PASS_Intelligence_SavageJansen2018'][1]\n",
    "df_res_Intelligence = scdrs.score_cell(adata, gene_list_Intelligence, gene_weight=gene_weight_Intelligence, n_ctrl=20)\n",
    "\n",
    "gene_list_risk = df_gs['PASS_GeneralRiskTolerance_KarlssonLinner2019'][0]\n",
    "gene_weight_risk = df_gs['PASS_GeneralRiskTolerance_KarlssonLinner2019'][1]\n",
    "df_res_risk = scdrs.score_cell(adata, gene_list_risk, gene_weight=gene_weight_risk, n_ctrl=20)\n",
    "\n",
    "gene_list_SCZ = df_gs['PASS_Schizophrenia_Pardinas2018'][0]\n",
    "gene_weight_SCZ = df_gs['PASS_Schizophrenia_Pardinas2018'][1]\n",
    "df_res_SCZ = scdrs.score_cell(adata, gene_list_SCZ, gene_weight=gene_weight_SCZ, n_ctrl=20)\n",
    "\n",
    "gene_list_BIP = df_gs['PASS_BIP_Stahl2019'][0]\n",
    "gene_weight_BIP = df_gs['PASS_BIP_Stahl2019'][1]\n",
    "df_res_BIP = scdrs.score_cell(adata, gene_list_BIP, gene_weight=gene_weight_BIP, n_ctrl=20)\n",
    "\n",
    "gene_list_ADHD = df_gs['PASS_ADHD_Demontis2018'][0]\n",
    "gene_weight_ADHD = df_gs['PASS_ADHD_Demontis2018'][1]\n",
    "df_res_ADHD = scdrs.score_cell(adata, gene_list_ADHD, gene_weight=gene_weight_ADHD, n_ctrl=20)\n",
    "\n",
    "adata.obs[\"norm_score_EDU_COLLEGE\"] = df_res_EDU_COLLEGE[\"norm_score\"]\n",
    "adata.obs[\"norm_score_EDU_YEARS\"] = df_res_EDU_YEARS[\"norm_score\"]\n",
    "adata.obs[\"norm_score_Verbal\"] = df_res_Verbal[\"norm_score\"]\n",
    "adata.obs[\"norm_score_Intelligence\"] = df_res_Intelligence[\"norm_score\"]\n",
    "adata.obs[\"norm_score_risk\"] = df_res_risk[\"norm_score\"]\n",
    "adata.obs[\"norm_score_SCZ\"] = df_res_SCZ[\"norm_score\"]\n",
    "adata.obs[\"norm_score_BIP\"] = df_res_BIP[\"norm_score\"]\n",
    "adata.obs[\"norm_score_ADHD\"] = df_res_ADHD[\"norm_score\"]\n",
    "adata.obs[\"norm_score_SWB\"] = df_res_SWB[\"norm_score\"]\n",
    "adata.obs[\"norm_score_MDD\"] = df_res_MDD[\"norm_score\"]\n",
    "adata.obs[\"norm_score_AD\"] = df_res_AD[\"norm_score\"]\n",
    "adata.obs[\"norm_score_Worry\"] = df_res_Worry[\"norm_score\"]\n",
    "\n",
    "sc.set_figure_params(figsize=[2.5, 2.5], dpi=150)\n",
    "scDRS=[\"norm_score_EDU_COLLEGE\", \"norm_score_EDU_YEARS\", \"norm_score_Verbal\", \"norm_score_Intelligence\",\n",
    "\"norm_score_risk\", \"norm_score_SCZ\", \"norm_score_BIP\", \"norm_score_ADHD\",\n",
    "\"norm_score_SWB\", \"norm_score_MDD\", \"norm_score_AD\",\"norm_score_Worry\"]\n",
    "\n",
    "sc.pl.umap(adata, color=scDRS, color_map=\"RdBu_r\", vmin=-5, vmax=5, s=20, save='all_scDRS.png')\n",
    "sc.pl.dotplot(adata, scDRS, groupby='leiden', use_raw=True, save='scDRS.png')\n",
    "sc.pl.dotplot(adata, scDRS, groupby='CellType', use_raw=True, save='celltypes-scDRS.png')\n",
    "\n",
    "\n",
    "#print(df_res.iloc[:4])\n",
    "#df_gs.to_csv(\"data/geneset.gs\", sep=\"\\t\")\n",
    "## List all keys\n",
    "#def getList(dict):\n",
    "#    return dict.keys()\n",
    "#print(getList(df_gs))\n",
    "#for trait in dict_score:\n",
    "#    adata.obs[trait] = dict_score[trait][\"norm_score\"]\n",
    "\n",
    "adata.write(results_file, compression='gzip')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4def6138",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Trajectory mapping\n",
    "\n",
    "sc.pl.paga(adata, color=['leiden'], save='tj.png')\n",
    "adata.uns['iroot'] = np.flatnonzero(adata.obs['leiden']  == '0')[0]\n",
    "sc.tl.dpt(adata)\n",
    "#sc.pl.draw_graph(adata, color=['leiden', 'dpt_pseudotime'], legend_loc='on data', save='pseudotime.png')\n",
    "sc.pl.umap(adata, color=['dpt_pseudotime'], save='pseudotime.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "192471f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Percent each cluster\n",
    "\n",
    "data = bc.tl.count_occurrence(adata, add_percentage=True, count_variable='leiden')\n",
    "display(data)\n",
    "fig = plt.figure(figsize=(15,6))\n",
    "#visualize distribution before logarithmization\n",
    "ax1 = fig.add_subplot(1, 2, 1)\n",
    "ax1 = sns.barplot(data = data, x = data.index.tolist(), y = 'Counts' )\n",
    "plt.xticks(rotation=90)\n",
    "#visualize distribution after logarithmization\n",
    "ax2 = fig.add_subplot(1, 2, 2)\n",
    "ax2 = sns.barplot(data = data, x = data.index.tolist(), y = 'Percentage' )\n",
    "plt.xticks(rotation=90)\n",
    "plt.savefig('percent_cluster.png', format='png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b612b51f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## Reassign CellType\n",
    "adata.obs['CellType'] = (\n",
    "    adata.obs[\"leiden\"]\n",
    "    .map(lambda x: {\n",
    "    '11': 'Excitatory 1', '13': 'Excitatory 1', '9': 'Excitatory 2',\n",
    "    '3': 'Neural progenitor 1', '5': 'Neural progenitor 1', '6': 'Neural progenitor 1', '16': 'Neural progenitor 1', \n",
    "    '15': 'Neural progenitor 2',\n",
    "    '0': 'Pluripotent 1', '12': 'Pluripotent 1', \n",
    "    '20': 'Pluripotent 2',\n",
    "    '18': 'Pluripotent 3', '19': 'Pluripotent 3', '21': 'Pluripotent 3',\n",
    "    '2': 'Pluripotent 4', '7': 'Pluripotent 4', '8': 'Pluripotent 4', \n",
    "    '4': 'Pluripotent 5', '14': 'Pluripotent 5', \n",
    "    '1': 'Pluripotent 6', '10': 'Pluripotent 6', '17': 'Pluripotent 6', \n",
    "    }.get(x, x)).astype(\"category\"))\n",
    "\n",
    "sc.pl.umap(adata, color=['CellType'], save='assigned_celltype.png')\n",
    "\n",
    "\n",
    "## cM graphs\n",
    "cmtx3 = sc.metrics.confusion_matrix(\"CellType\", \"leiden\", adata.obs)\n",
    "heatmap_plot4 = sns.heatmap(cmtx3)\n",
    "fig4 = heatmap_plot4.get_figure()\n",
    "fig4.savefig(\"celltype_cM.png\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "219774a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## CellType Marker genes, wilcoxon\n",
    "sc.tl.rank_genes_groups(adata, 'CellType', method='wilcoxon')\n",
    "sc.pl.rank_genes_groups_dotplot(adata, n_genes=4, save='celltype_mg_dotplot.png')\n",
    "sc.pl.rank_genes_groups_matrixplot(adata, n_genes=3, standard_scale='var', cmap='Blues', save='celltype_wilcoxon_mg_matrixplot.png')\n",
    "sc.pl.rank_genes_groups_dotplot(adata, n_genes=3, standard_scale='var', cmap='Blues', save='celltype_wilcoxon_dotplot.png')\n",
    "adata.write(results_file)\n",
    "DEcelltype=pd.DataFrame(adata.uns['rank_genes_groups']['names']).head(25)\n",
    "DEcelltype.columns ='leiden_' + DEcelltype.columns\n",
    "DEcelltype.to_csv(ctwilcm)\n",
    "\n",
    "## Percent each cell type\n",
    "data = bc.tl.count_occurrence(adata, add_percentage=True, count_variable='CellType')\n",
    "display(data)\n",
    "fig = plt.figure(figsize=(15,6))\n",
    "#visualize distribution before logarithmization\n",
    "ax1 = fig.add_subplot(1, 2, 1)\n",
    "ax1 = sns.barplot(data = data, x = data.index.tolist(), y = 'Counts' )\n",
    "plt.xticks(rotation=25)\n",
    "#visualize distribution after logarithmization\n",
    "ax2 = fig.add_subplot(1, 2, 2)\n",
    "ax2 = sns.barplot(data = data, x = data.index.tolist(), y = 'Percentage' )\n",
    "plt.xticks(rotation=25)\n",
    "plt.savefig('percent_celltype.png', format='png')\n",
    "\n",
    "sc.pl.dotplot(adata, scDRS, groupby='CellType', use_raw=True, save='celltypes-scDRS.png')\n",
    "adata.write(results_file, compression='gzip')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rxrx",
   "language": "python",
   "name": "rxrx"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
